### PR TITLE
merge safari into develop

### PR DESCRIPTION
**Pull request by jdorweiler**
**Reviewer:**

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/html/background.html [hunk 1]
```diff
@@ -17,10 +17,7 @@
 <!DOCTYPE html>
 <html>
   <head>
-    <script src="../public/background.js"></script>
-    <script type="text/javascript">
-      var background = new Background();
-    </script>
```
Comment (andrey-p) [inline shared/html/background.html]: Wow, didn't even realize this was happening.

File: shared/js/background/abp-lists.es6.js [hunk 2]
```diff
@@ -12,6 +12,7 @@ const surrogates = require('./surrogates.es6')
 const settings = require('./settings.es6')
 const ATB = require('./atb.es6')
 const load = require('./load.es6')
+const polyfill = require('./safari-polyfill.es6')
```
Comment (andrey-p) [inline shared/js/background/abp-lists.es6.js]: I've been thinking about the "polyfill" name. I'm not sure "polyfill" is quite there. You could say the Safari is kind of a polyfill for features that the webextensions API normally supports, but in that case the name doesn't work for Chrome and FF.

How about something like "[chrome/safari]-wrapper"? "[chrome/safari]-extension-utils"? Or just "[chrome/safari]-utils"?

File: shared/js/background/tracker-lists.es6.js [hunk 3]
```diff
@@ -14,15 +13,19 @@ function setList (name, data) {
...
             console.log(`Loaded tracker list: ${listLocation}/${listName}`)
             lists[listName.replace('.json', '')] = listJSON
         })
     })
 }
 
-settings.ready().then(() => loadLists())
+loadLists()
 
 module.exports = {
```
Comment (andrey-p) [inline shared/js/background/tracker-lists.es6.js]: This whole thing simply loads a local JSON file. Let's just replace it with a `require()`.

File: shared/js/background/utils.es6.js [hunk 4]
```diff
@@ -81,6 +80,14 @@ function getBrowserName() {
     return browser
 }
 
+function setBadgeIcon (path, target) {
+    polyfill.setBadgeIcon(path, target)
+}
+
+function getExtensionURL (path) {
+    return polyfill.getExtensionURL(path)
+}
```
Comment (andrey-p) [inline shared/js/background/utils.es6.js]: Why do these need to go through `utils`?

File: shared/js/background/tracker-lists.es6.js (hunk 3 above)
Comment (jdorweiler) [reply shared/js/background/tracker-lists.es6.js]: Yeah good idea. That would make my life easier too.

File: shared/scss/popup.scss [hunk 5]
```diff
@@ -285,6 +296,9 @@ body {
                             background-image: url($icon__trackers__unblocked--moz);
                         }
 
+                        .is-browser--safari & {
+                            background-image: url($icon__trackers__unblocked--safari);
+                        }
```
Comment (andrey-p) [inline shared/scss/popup.scss]: I'm just going to update everything to use the `cross_platform_bg_image` mixin to make this easier. Lemme do that in a separate PR.

**Comment by jdorweiler** [comment]
@andrey-p Ready to go, finally. 
Build with `npm run dev-safari`
Load in extension builder from `build/duckduckgo.safariextension/dev/duckduckgo.safariextension`

**Comment by andrey-p** [comment]
Awesome! Looking at it now.

File: shared/js/background/classes/site.es6.js [hunk 6]
```diff
@@ -86,7 +86,7 @@ class Site {
         if (this.domain === 'extensions')
             return 'extensions'
 
-        if (this.domain === chrome.runtime.id)
+        if (window.chrome && this.domain === chrome.runtime.id)
             return 'options'
```
Comment (andrey-p) [inline shared/js/background/classes/site.es6.js]: Not specifically related to this change, but the popup looks broken when you've opened e.g. the Options page. The connection string doesn't show, and it says "undefined trackers blocked".

File: shared/js/ui/views/mixins/open-options-page.es6.js [hunk 7]
```diff
@@ -1,12 +1,21 @@
...
+      } 
+      else if (browser === 'chrome') {
         window.chrome.runtime.openOptionsPage()
       }
+      else if (browser === 'safari') {
+          let tab = safari.application.activeBrowserWindow.openTab()
+          tab.url = browserWrapper.getExtensionURL('html/options.html')
+          // close the popup
+          safari.self.hide()
+      }
```
Comment (andrey-p) [inline shared/js/ui/views/mixins/open-options-page.es6.js]: [suggestion] Worth doing something like `browserUIWrapper.openOptionsPage()` instead?

File: shared/js/background/safari-events.es6.js [hunk 8]
```diff
@@ -0,0 +1,295 @@
+const ATB = require('./atb.es6')
+const tabManager = require('./tab-manager.es6')
+const utils = require('./utils.es6')
+const Companies = require('./companies.es6')
+const settings = require('./settings.es6')
+const abpLists = require('./abp-lists.es6')
+
+// add these to contentWindow so the popup can access them
+safari.extension.globalPage.contentWindow.tabManager = tabManager
+safari.extension.globalPage.contentWindow.Companies = Companies
```
Comment (andrey-p) [inline shared/js/background/safari-events.es6.js]: Is it not possible to do messages between the popup and the background processes? I know they run in the same context, but it would make things loads cleaner, and would probably help us standardize between the Chrome/FF and Safari a bit more.

File: shared/js/ui/views/search.es6.js [hunk 9]
```diff
@@ -39,7 +39,11 @@ Search.prototype = window.$.extend({},
 
     _handleBurgerClick: function (e) {
       e.preventDefault()
-      this.model.send('burgerClick')
+      if (window.safari && this.model.searchText) {
+          this._handleSubmit(e)
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply shared/js/ui/views/search.es6.js]: Could be a bug. For some reason the form submit runs this handler in safari.

File: shared/js/background/safari-events.es6.js (hunk 8 above)
Comment (jdorweiler) [reply shared/js/background/safari-events.es6.js]: Nope we can't message the background. That's one of the big complications with Safari.

File: shared/js/background/safari-events.es6.js [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+ * Safari doesn't have a onInstalled event so we'll set a flag in localStorage.
+ * 1. check installed flag
+ * 2. go through all current tabs and recreate our own internal tab objects
+ * 3. open post install page only if we're on a DDG page or Safari gallery page
+ */
+let onStartup = (() => {
+    if (!localStorage['installed']) {
+        localStorage['installed'] = true
+        ATB.updateATBValues()
+    }
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply shared/js/background/safari-events.es6.js]: Yeah that was all tested in the current Safari extension. Everyone should be upgraded by now but it's probably safe to just leave this here.

File: shared/js/background/safari-events.es6.js (hunk 8 above)
Comment (andrey-p) [reply shared/js/background/safari-events.es6.js]: Potentially stupid idea, what if we faked a messaging interface? E.g. off the top of my head:

```js
// in UI code
safari.extension.globalPage.contentWindow.sendMessage({ getSetting: 'foo' })
   .then(() => { bar() })

// in background code
safari.extension.globalPage.contentWindow.sendMessage = (message) => {
    if (message.getSetting) {
        return Promise.resolve(valueFromSettings)
    }
}
```

The benefit will be that at least Safari and Chrome/FF will communicate in a similar way.

File: shared/js/ui/views/search.es6.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (andrey-p) [reply shared/js/ui/views/search.es6.js]: I wonder if Safari does something weird because it's a `<button>` inside a `<form>`?

Maybe we can just move the button out of the form (https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/2a3b3a33128d34ed044d292fee5a70f77c086104/shared/js/ui/templates/search.es6.js#L13) or try and change it from a `<button>` to a `<a>`.

If something like that works I'd rather do that than leaving the click handler in a hacky place.

File: shared/js/background/safari-events.es6.js (hunk 8 above)
Comment (jdorweiler) [reply shared/js/background/safari-events.es6.js]: It's a good idea. Let's save that for a follow up. I'd want to spend some time cleaning up `fetch` to work with that too.

File: shared/js/background/classes/site.es6.js (hunk 6 above)
Comment (jdorweiler) [reply shared/js/background/classes/site.es6.js]: Thanks, fixed that elsewhere.

File: shared/js/background/tab-manager.es6.js [hunk 11]
```diff
@@ -1,22 +1,67 @@
...
 class TabManager {
     constructor() {
         this.tabContainer = {}
     };
 
+    /* Get stashed tabId from native safari tabs. This needs to 
+     * be here for now. The popup can't access the tabId for
+     * some reason
+     */
+    getTabId(e) {
```
Comment (jdorweiler) [inline shared/js/background/tab-manager.es6.js]: These will need to stay here for now. There's some bug with accessing the active safari tab from inside the popup that makes no sense to me.

File: shared/js/background/trackers.es6.js [hunk 12]
```diff
@@ -298,5 +301,6 @@ function checkABPParsedList(list, url, siteDomain, request) {
 }
 
 module.exports = {
-    isTracker: isTracker
+    isTracker: isTracker,
+    loadLists: loadLists
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply shared/js/background/trackers.es6.js]: Let's fix later with the other loadLists in tracker-lists. I had some timing issues with the entityMap that this seemed to fix.

File: shared/js/ui/views/search.es6.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jdorweiler) [reply shared/js/ui/views/search.es6.js]: I didn't get to this one yet. I'll check it out tomorrow.

File: shared/js/background/trackers.es6.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (andrey-p) [reply shared/js/background/trackers.es6.js]: Cool. I think some of our work is cut out over at https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/196.

File: shared/js/ui/templates/shared/hamburger-button.es6.js [hunk 13]
```diff
@@ -2,7 +2,7 @@ const bel = require('bel')
 
 module.exports = function (klass) {
   klass = klass || ''
-  return bel`<button class="hamburger-button ${klass}">
+  return bel`<button type="button" class="hamburger-button ${klass}">
```
Comment (andrey-p) [inline shared/js/ui/templates/shared/hamburger-button.es6.js]: cc @jdorweiler this and https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/195/files#diff-6a64f88be34c8f8036a1faf3c3192e6bR35 fix the hamburger menu problem.

File: shared/js/ui/base/chrome-ui-wrapper.es6.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+let getBackgroundTabData = ((thisModel) => {
+    return new Promise((resolve, reject) => {
+        thisModel.fetch({getCurrentTab: true}).then((tab) => {
+            if (tab) {
+                thisModel.fetch({getTab: tab.id}).then((backgroundTabObj) => {
+                    resolve(backgroundTabObj)
+                })
+            }
+        })
+    })
```
Comment (andrey-p) [inline shared/js/ui/base/chrome-ui-wrapper.es6.js]: [nitpick] why does this need to go via `thisModel`? Seems you can just use the `fetch` that's within this module.

File: shared/scss/_mixins.scss [hunk 15]
```diff
@@ -69,7 +69,8 @@
         position: absolute;
         right: 14px;
         width: 32px;
-        height: auto;
+        height: 22px;
+        margin-top: -1px;
```
Comment (andrey-p) [inline shared/scss/_mixins.scss]: This makes the toggle appear vertically off-centre.

Try setting `top: 16px` on `.site-info__toggle-container` in `popup.scss` instead of doing the margin thing.

I have no idea why the toggle gets cut off without this. I played around with this and I'm stumped.

**Comment by andrey-p** [comment]
Spotted another issue:

1. Open any page e.g. theguardian.com
2. Open the popup before the site is finished loading
3. Safari shows "D upgraded to C", which changes to "C upgraded to D" when the site finishes loading - seems this is being shown instead of the "calculating" state.

If you do it with duckduckgo.com, it shows the "new tab" state instead.

File: shared/scss/_mixins.scss (hunk 15 above)
Comment (MariagraziaAlastra) [reply shared/scss/_mixins.scss]: @andrey-p I assume you're testing in Safari? I see the toggle vertically centered with the current styles. This was a fix for the cut off knob.
The weird thing is there was already a fix for it in the past, see [here](https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/51), but it occurred again while porting the latest changes.

Comment (andrey-p) [reply shared/scss/_mixins.scss]: @MariagraziaAlastra I'm seeing this in Chrome/FF. It's pretty noticeable for me.

screenshot from 2018-04-12 14-46-00

Comment (andrey-p) [reply shared/scss/_mixins.scss]: That fix also looks like it could do the job - it didn't make it into this PR since it was merged into the old branch, I guess.

Comment (MariagraziaAlastra) [reply shared/scss/_mixins.scss]: @andrey-p no, it's just the old fix didn't work anymore in this case :(

Comment (MariagraziaAlastra) [reply shared/scss/_mixins.scss]: Anyway thanks! The cut off knob fix in this PR is actually mine - Tried to fix it in the Safari browser inspector - so I take the blame for it messing with the vertical alignment in FF/Chrome 😄

File: shared/js/ui/templates/shared/hamburger-button.es6.js (hunk 13 above)
Comment (jdorweiler) [reply shared/js/ui/templates/shared/hamburger-button.es6.js]: 👍 thanks

File: shared/js/ui/base/chrome-ui-wrapper.es6.js (hunk 14 above)
Comment (jdorweiler) [reply shared/js/ui/base/chrome-ui-wrapper.es6.js]: 👍

**Comment by jdorweiler** [comment]
Ok ready to go again. 
I never got the calculating state working correctly in the other Safari extension. I'll make a follow up to look into it more.

File: shared/scss/_mixins.scss (hunk 15 above)
Comment (jdorweiler) [reply shared/scss/_mixins.scss]: Changing the margin on `.site-info__toggle-container` doesn't fix it.

**Comment by MariagraziaAlastra** [comment]
Went through the whole manual regression test suite in Safari, Chrome and FF.
No regressions in Chrome and FF, looks mergeable for Safari too 👍

**Comment by andrey-p** [comment]
I think I got the button fixed. Let's get the beta tests done and I think this'll be ready to merge :tada:

**Comment by andrey-p** [comment]
:+1:
